backend/tilavarauspalvelu/models/payment_order/actions.py

```diff
@@ -29,73 +33,92 @@
-    def get_order_status_from_webshop_payment(self, webshop_payment: Payment | None) -> OrderStatus:
-        """Determines the order status based on the payment response from the webshop."""
-        order_status = OrderStatus(self.payment_order.status)
-        if order_status in OrderStatus.finalized:
-            return order_status
+        webshop_payment = VerkkokauppaAPIClient.get_payment(order_uuid=self.payment_order.remote_id)
 
         if webshop_payment is None:
-            # User has not entered payment phase in webshop within the expiration time
-            if self.payment_order.expires_at > local_datetime():
-                return OrderStatus.DRAFT
-            return OrderStatus.EXPIRED
+            new_status = self.get_order_status_if_no_webshop_payment()
+            payment_id: str = ""
+        else:
+            new_status = self.get_order_status_from_webshop_payment(webshop_payment)
+            payment_id = webshop_payment.payment_id
 
-        older_than_minutes = settings.VERKKOKAUPPA_ORDER_EXPIRATION_MINUTES
-        webshop_payment_expires_at = local_datetime() - datetime.timedelta(minutes=older_than_minutes)
+        if new_status == self.payment_order.status:
+            return
+
+        self.payment_order.status = new_status
+        self.payment_order.processed_at = local_datetime()
+        self.payment_order.payment_id = payment_id
+        self.payment_order.save(update_fields=["status", "processed_at", "payment_id"])
+
+        self.complete_payment()
+
+    def get_order_status_from_webshop_payment(self, webshop_payment: Payment) -> OrderStatus:
+        is_handled_payment = self.payment_order.is_handled_payment
+        is_handled_payment_overdue = self.payment_order.is_overdue_handled_payment
+        is_invoiced = webshop_payment.payment_gateway == WebShopPaymentGateway.INVOICE
+        webshop_payment_expiration_minutes = datetime.timedelta(minutes=settings.VERKKOKAUPPA_ORDER_EXPIRATION_MINUTES)
+        webshop_payment_expires_at = webshop_payment.timestamp + webshop_payment_expiration_minutes
+        webshop_payment_expired = webshop_payment_expires_at <= local_datetime()
 
         match webshop_payment.status:
             case WebShopPaymentStatus.CANCELLED:
                 return OrderStatus.CANCELLED
 
-            case WebShopPaymentStatus.PAID_ONLINE if webshop_payment.payment_gateway == WebShopPaymentGateway.INVOICE:
+            case WebShopPaymentStatus.PAID_ONLINE if is_invoiced:
                 return OrderStatus.PAID_BY_INVOICE
 
             case WebShopPaymentStatus.PAID_ONLINE:
                 return OrderStatus.PAID
 
+            # Handled reservation payments do not expire even if the payment expires.
+            case WebShopPaymentStatus.CREATED if is_handled_payment:
+                return OrderStatus.PENDING
+
+            # Handled reservation payments expire immediately if they are overdue.
+            case WebShopPaymentStatus.CREATED if is_handled_payment_overdue:
```

Comment:
Should check for overdue before just generally handled.

Reply:
Fixed